This is in-house developed patch reported at
https://rt.perl.org/Ticket/Display.html?id=121972

Expectation is that future perl versions will contain the fix.
--- perl-5.16.3/util.c	2014-05-28 10:46:26.254431769 +0200
+++ perl-5.16.3/util.c	2014-05-28 10:45:56.837881294 +0200
@@ -2092,7 +2092,7 @@ Perl_my_setenv(pTHX_ const char *nam, co
        my_setenv_format(environ[i], nam, nlen, val, vlen);
     } else {
 # endif
-#   if defined(__CYGWIN__) || defined(EPOC) || defined(__SYMBIAN32__) || defined(__riscos__)
+#   if defined(__CYGWIN__) || defined(EPOC) || defined(__SYMBIAN32__) || defined(__riscos__) || defined(__sun)
 #       if defined(HAS_UNSETENV)
         if (val == NULL) {
             (void)unsetenv(nam);
